### PR TITLE
Webserver

### PR DESCRIPTION
**Pull request by Noordfrees**
Closes https://github.com/widelands/widelands-metaserver/issues/71.

I'm hoping someone will review this, with especial regard to security.

This provides an add-ons web server. How to set up, configure and run your own local server is explained in the README. (yes the first-time setup is sort of involved…). It can be tested together with the https://github.com/Noordfrees/widelands/tree/addons-webserver branch (which conveniently tries to connect to a server on localhost by default currently).

The process to get the server ready is to first merge this PR; then (at any later time) configure it on a real server; finally when that is functional and has been well tested, merge the main repo branch (https://github.com/widelands/widelands/pull/4934).

Things that were not tested: How it performs when a hundred Widelands instances access the server at once (I only ensured that 2 clients don't block each other – this will surely need some follow-up branches when it's in production); a really determined hacker's attack (all the easy things are caught and rejected though). I also don't know how often the server will request a maintainer's attention initially but I guess it'll keep me busy while it's new. This *is* the first server I've written after all :smile:

### CODE REVIEW

**Comment by matthiakl** [comment]
Looks promising :rocket: 
I have set up the server locally and will start testing it. During the process I already have some general remarks:
- As there is now more java code in the repo, it would be best to move it to a separate folder. Otherwise the toplevel directory is cluttered with all the .class files.
- clang-format also works for java. How about using the same github action as for the main repo and just set the option `Language Java`
- The Server.java is very long and hard to read. Additionally it contains multiple classes and could easily be split into multiple files. Also splitting up the CMD_* implementations into methods improves readability.
- The config file parser does not strip spaces around the `=` and throws an exception on a line like `port = 1234`
- The add-ons are still served file by file. Do you already plan serving add-ons as zip? The server could pack the most recent version of each add-on as zip after during the github sync and serve an add-on as on archive, which is then extracted by the client. (Just an idea for the future, not for now)

**Comment by Noordfrees** [comment]
All fixed except the last one (clang-format action still needs some tweaking). 
I had initially (= before starting on an actual server) considered storing the add-ons as zip files only and letting Widelands extract them after downloading, but it turned out that our zip filesystem implementation does not allow doing this easily. So if someone wants to code this (or switch to a better library, see https://github.com/widelands/widelands/issues/399) I'd not be opposed. However in my tests the time required to read a large add-on and write it to the network is far less than the time required by Widelands to dump the streams to files, so I'm not sure we would even gain a significant performance bonus by transferring a zip.

**Comment by matthiakl** [comment]
Wow, the download speed is really FAST :+1: 
I successfully created a user and logged in (as admin), but if I try to upload an add-on (existing or new), I get a Java exception without further message:
```
[00:28:11.194 real] ERROR: upload addon Tournament_2021_win_condition.wad: Expected end of stream, received:
[00:28:11.194 real] ERROR: java.lang.reflect.InvocationTargetException
```
The files are created in the servers temp dir, but not transferred to the addons dir, and deleted after the above error.

**Comment by Noordfrees** [comment]
The most recent refactoring obscured the error message because of some internal Java magic. Now you should see the actual error instead.

**Comment by frankystone** [comment]
Are there any ideas on which webserver this will be installed/running?

I can check if our server has the capabilities and give Noordfrees access to it. But it would be good to have stonerl available when installing the addons server. He is very familiar with configuring the systemd scripts ... but unfortunately he is very busy for month now :-(

**Comment by Noordfrees** [comment]
I've spoken about this with SirVer already and we've agreed that the add-ons webserver will run in a docker container on the server. I've tried out locally how to get the webserver running that way. Someone who knows the server setup would be good though… Do you know if the server already has docker set up or will this need to be newly installed/configured first?

**Comment by frankystone** [comment]
Ah, and stonerl might be, or not, available on IRC in the german channel #widelands-de (either on freenode or liberachat).

**Comment by Noordfrees** [comment]
I just wanted to mention that a review of this branch would be much appreciated :) 
I'm ready to deploy as soon as this is merged, and it would be good to be able to test the server in action, and I'll have lots of time in August to deal with any initial problems that can be expected while the add-ons server is still new. There are also a few features and refactorings that should be added soonish but would exceed the bounds of this PR.

File: wl/server/Server.java
Comment (matthiakl) [inline wl/server/Server.java:95]: Consider using a log function, which adds the timestamp prefix.
```
void log(String log) { 
    System.out.println("[" + new Date() + " @ " + Thread.currentThread().getName() + "] " + log);
}
```

File: wl/server/Server.java
Comment (matthiakl) [inline wl/server/Server.java:89]: If you implement the `run` methods in distinct classes the `main` function becomes much more readable
```
class SyncThread implements Runnable {
... the github sync run() method ...
}
class ConnectionThread implements Runnable {
... this run() method ...
}
...
new Thread(new SyncThread()).start();
new Thread(new ConnectionThread()).start();
```
Then it is also easier to pull out some of the run-code into methods.

**Review by matthiakl — COMMENTED** [review]
I was quite busy the last weeks, but will try to finish testing over the weekend. Meanwhile I looked over the code, which looks good, but is still sometimes very long.

I will report the client side issues I encounterd in the PR of the main repo.

**Comment by matthiakl** [comment]
Somehow I managed to get -1 votes into the metadata file of an add-on. This should be cut to 0.
If the SQL connection breaks while the server is running, there is no recovery, but since that requires a maintainers attention anyway it should be fine for now.
The character encoding does work correctly now.
Now, if I upload an add-on I get `ERROR: java.lang.NullPointerException: Cannot read field "value" because the return value of "java.util.TreeMap.get(Object)" is null`.  Afterwards this error also appears in the contact form.

**Comment by Noordfrees** [comment]
Fixed by recalculating the votes count from the uservotes file after every change, instead of assuming the counts always match up. I cannot reproduce the nullpointer error though – perhaps the add-on's config file is invalid?

**Comment by matthiakl** [comment]
I cannot reproduce the error with a clean homedir, so it was probably a faulty file.
The server side looks good now. To improve robustness and security any size parameter in a clients command should be checked to have a reasonable size. E.g. imagine a submit command with 65000 directories each with 65000 files.
I will try to test the connection from Windows now.

**Review by matthiakl — APPROVED** [review]
I'm happy with this part now :)

**Comment by Noordfrees** [comment]
Thanks for the review! :)
Will deploy this now

**Pull request merged**
